framework/src/org/checkerframework/framework/util/LubVisitor.java

```diff
@@ -0,0 +1,350 @@
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedUnionType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedWildcardType;
+import org.checkerframework.framework.type.QualifierHierarchy;
+import org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor;
+import org.checkerframework.framework.type.visitor.VisitHistory;
+import org.checkerframework.javacutil.ErrorReporter;
+import org.checkerframework.javacutil.InternalUtils;
+import org.checkerframework.javacutil.TypesUtils;
+
+public class LubVisitor extends AbstractAtmComboVisitor<Void, AnnotatedTypeMirror> {
+
+    public AnnotatedTypeFactory atypeFactory;
+    public QualifierHierarchy qualifierHierarchy;
+    public VisitHistory visitHistory = new VisitHistory();
+
+    public LubVisitor(AnnotatedTypeFactory atypeFactory) {
+        this.atypeFactory = atypeFactory;
+        this.qualifierHierarchy = atypeFactory.getQualifierHierarchy();
+    }
+
+    /**
+     * Returns an ATM that is the least upper bound of type1 and type2 and whose Java type is
+     * lubJavaType.  lubJavaType must be a super type or convertible to the Java types of type1 and
+     * type2.
+     **/
+    AnnotatedTypeMirror lub(
+            AnnotatedTypeMirror type1, AnnotatedTypeMirror type2, TypeMirror lubJavaType) {
+        AnnotatedTypeMirror lub = AnnotatedTypeMirror.createType(lubJavaType, atypeFactory, false);
+
+        if (type1.getKind() == TypeKind.NULL) {
+            return lubWithNull((AnnotatedNullType) type1, type2, lub);
+        }
+        if (type2.getKind() == TypeKind.NULL) {
+            return lubWithNull((AnnotatedNullType) type2, type1, lub);
+        }
+
+        AnnotatedTypeMirror type1AsLub = AnnotatedTypes.asSuper(atypeFactory, type1, lub);
+        AnnotatedTypeMirror type2AsLub = AnnotatedTypes.asSuper(atypeFactory, type2, lub);
+
+        visit(type1AsLub, type2AsLub, lub);
+        return lub;
+    }
+
+    /**
+     * Returns the leastUpperBound of an AnnotatedNullType and another AnnotatedTypeMirror.
+     *
+     * @param nullType  type to lub
+     * @param otherType other type to lub
+     * @param lub       type of lub
+     * @return the lub of {@code nullType} and {@code otherType}
+     */
+    private AnnotatedTypeMirror lubWithNull(
+            AnnotatedNullType nullType, AnnotatedTypeMirror otherType, AnnotatedTypeMirror lub) {
+
+        AnnotatedTypeMirror otherAsLub = AnnotatedTypes.asSuper(atypeFactory, otherType, lub);
+
+        if (otherAsLub.getKind() != TypeKind.TYPEVAR && otherAsLub.getKind() != TypeKind.WILDCARD) {
+            for (AnnotationMirror nullAnno : nullType.getAnnotations()) {
+                AnnotationMirror otherAnno = otherAsLub.getAnnotationInHierarchy(nullAnno);
+                AnnotationMirror lubAnno = qualifierHierarchy.leastUpperBound(nullAnno, otherAnno);
+                lub.replaceAnnotation(lubAnno);
```

Comment:
Note that this modifies the input parameter lub. This should be mentioned in the javadoc of this method and callers.